crates/bevy_ecs/src/world/entity_ref.rs

```diff
@@ -0,0 +1,889 @@
+                archetype.set_entity_table_row(swapped_location.index, old_table_row);
+            }
+
+            new_location
+        };
+
+        self.location = new_location;
+        entities.meta[self.entity.id as usize].location = new_location;
+    }
+
+    pub fn insert<T: Component>(&mut self, value: T) -> &mut Self {
+        self.insert_bundle((value,))
+    }
+
+    pub fn remove<T: Component>(&mut self) -> Option<T> {
+        self.remove_bundle::<(T,)>().map(|v| v.0)
+    }
+
+    pub fn despawn(self) {
+        let world = self.world;
+        world.flush();
+        let location = world
+            .entities
+            .free(self.entity)
+            .expect("entity should exist at this point.");
+        let table_row;
+        let moved_entity;
+        {
+            // SAFE: entity is live and is contained in an archetype that exists
+            let archetype = unsafe { world.archetypes.get_unchecked_mut(location.archetype_id) };
+            for component_id in archetype.components() {
+                let removed_components = world
+                    .removed_components
+                    .get_or_insert_with(component_id, Vec::new);
+                removed_components.push(self.entity);
+            }
+            let remove_result = archetype.swap_remove(location.index);
+            if let Some(swapped_entity) = remove_result.swapped_entity {
+                world.entities.meta[swapped_entity.id as usize].location = location;
+            }
+            table_row = remove_result.table_row;
+
+            for component_id in archetype.sparse_set_components() {
+                let sparse_set = world.storages.sparse_sets.get_mut(*component_id).unwrap();
+                sparse_set.remove(self.entity);
+            }
+            // SAFE: tables and table rows stored in archetypes always exist
+            moved_entity = unsafe {
+                world
+                    .storages
+                    .tables
+                    .get_unchecked_mut(archetype.table_id())
+                    .swap_remove_unchecked(table_row)
+            };
+        };
+
+        if let Some(moved_entity) = moved_entity {
+            let moved_location = world.entities.get(moved_entity).unwrap();
+            // SAFE: entity is live and is contained in an archetype that exists
+            let archetype = unsafe {
+                world
+                    .archetypes
+                    .get_unchecked_mut(moved_location.archetype_id)
+            };
+            archetype.set_entity_table_row(moved_location.index, table_row);
+        }
+    }
+
+    /// # Safety
+    /// Caller should not modify the world in a way that changes the current entity's location
+    /// If the caller _does_ do something that could change the location, self.update_location() should be
+    /// called before using any other methods in EntityMut
+    pub unsafe fn world(&mut self) -> &mut World {
+        self.world
+    }
```

Comment:
```suggestion
    pub fn world(&mut self) -> &World {
        self.world
    }
    /// # Safety
    /// Caller should not modify the world in a way that changes the current entity's location
    /// If the caller _does_ do something that could change the location, self.update_location() must be
    /// called before using any other methods in EntityMut
    pub unsafe fn world_mut(&mut self) -> &mut World {
        self.world
    }
```
* split `world` and `world_mut`
* s/should/must